### PR TITLE
fix: improve OpenAI-compatible error handling and add test coverage

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Improves error handling reliability for OpenAI-compatible providers and adds comprehensive test coverage.

I tried using Goose through a gateway which had an incorrect max payload size. The gateway returned a plain text error.

Before, the underlying error was obfuscated by a JSON failure: `Error: Request failed: error decoding response body`
```bash
$ goose run --debug -t 'should I install goose?'
starting session | provider: openai model: qwen3
    logging to /Users/codefromthecrypt/.local/share/goose/sessions/20250819_090611.jsonl
    working directory: /Users/codefromthecrypt/oss/goose
Error: Request failed: error decoding response body
Interrupted before the model replied and removed the last message.
  error: The error above was an exception we were not able to handle.
These errors are often related to connection or authentication
We've removed the conversation up to the most recent user message
- depending on the error you may be able to continue
```

Now, the error is easier to see:
`Error: Request failed: Request failed with status: 413 Payload Too Large`
```bash
$  ./target/debug/goose run --debug -t 'should I install goose?'
starting session | provider: openai model: qwen3
    logging to /Users/codefromthecrypt/.local/share/goose/sessions/20250819_090540.jsonl
    working directory: /Users/codefromthecrypt/oss/goose
  2025-08-19T02:05:53.160383Z  WARN goose::providers::utils: Provider request failed with status: 413 Payload Too Large. Payload: None. Returning error: RequestFailed("Request failed with status: 413 Payload Too Large")
    at crates/goose/src/providers/utils.rs:113

Error: Request failed: Request failed with status: 413 Payload Too Large
Interrupted before the model replied and removed the last message.
  error: The error above was an exception we were not able to handle.
These errors are often related to connection or authentication
We've removed the conversation up to the most recent user message
- depending on the error you may be able to continue
```

**Changes:**
- Add `PartialEq` derive to `ProviderError` enum to enable testing
- Refactor `handle_status_openai_compat` to improve error parsing and reduce complexity
- Enhance authentication error message formatting with cleaner conditional logic
- Add comprehensive test suite covering various error scenarios and edge cases
- Improve function documentation with clear references to OpenAI error codes

**Impact:**
- More reliable error handling when provider APIs return malformed responses
- Better error messages for authentication failures
- Increased test coverage prevents regressions in error handling logic
- Cleaner code structure improves maintainability

### CODE REVIEW

**Comment by michaelneale** [comment]
thanks Adrian tagging @DOsinga as I know he had designs on making openai and providers more declarative and configurable

will need a good old cargo fmt run of course

File: crates/goose/src/providers/utils.rs [hunk 1]
```diff
@@ -118,45 +121,52 @@ pub fn map_http_error_to_provider_error(
...
+                ));
+            } else {
+                return Err(ProviderError::RequestFailed(format!(
+                    "{} (status {})",
+                    err,
+                    status.as_u16()
+                )));
             }
         }
+        // Fall through if not a valid OpenAI error response
```
Comment (DOsinga) [inline crates/goose/src/providers/utils.rs]: remove the comment - the rest are still valid OpenAI error responses, just not specific

Comment (codefromthecrypt) [reply crates/goose/src/providers/utils.rs]: ack!

**Comment by codefromthecrypt** [comment]
rebased!

**Comment by michaelneale** [comment]
running now...

**Review by michaelneale — APPROVED** [review]

File: crates/goose/src/providers/utils.rs [hunk 2]
```diff
@@ -243,8 +255,8 @@ pub async fn handle_response_google_compat(response: Response) -> Result<Value,
             );
             Err(ProviderError::RequestFailed(format!("Request failed with status: {}. Message: {}", final_status, error_msg)))
         }
-        _ if final_status.is_server_error() => {
-            Err(ProviderError::ServerError(format!("{:?}", payload)))
+        StatusCode::TOO_MANY_REQUESTS => {
+            Err(ProviderError::RateLimitExceeded(format!("{:?}", payload)))
```
Comment (jamadeo) [inline crates/goose/src/providers/utils.rs]: @DOsinga double check this merged correctly with #4160 ?

Comment (DOsinga) [reply crates/goose/src/providers/utils.rs]: oh I imagined this was a false positive and that it had to do with showing a weird diff here because of fork? can you check @codefromthecrypt ? we do want to retry all 500s now

Comment (michaelneale) [reply crates/goose/src/providers/utils.rs]: yeah I think this could go back to the other way, I have seen many 500s which are retry-able

Comment (codefromthecrypt) [reply crates/goose/src/providers/utils.rs]: whoops sorry, lemme sort this

Comment (codefromthecrypt) [reply crates/goose/src/providers/utils.rs]: ok re-open the chat if I didn't get it right this time!

File: crates/goose/src/providers/utils.rs [hunk 3]
```diff
@@ -243,10 +253,10 @@ pub async fn handle_response_google_compat(response: Response) -> Result<Value,
...
         }
-        StatusCode::INTERNAL_SERVER_ERROR | StatusCode::SERVICE_UNAVAILABLE => {
+        _ if final_status.is_server_error() => {
```
Comment (michaelneale) [inline crates/goose/src/providers/utils.rs]: ok so this will be a catch all - and even if TOO_MANY_REQUESTS triggers above it will retry?

Comment (codefromthecrypt) [reply crates/goose/src/providers/utils.rs]: I just reverted the whole thing. sorry for stealing your attention to an unrelated area of code.

**Comment by michaelneale** [comment]
get it in when the lights are green!

File: crates/goose/src/providers/utils.rs (hunk 3 above)
Comment (michaelneale) [reply crates/goose/src/providers/utils.rs]: @codefromthecrypt You're absolutely right!

**Comment by codefromthecrypt** [comment]
I see 💚
